harvester/wacz.py

```diff
@@ -0,0 +1,273 @@
+    a WACZ archive, and some utilities to actually extract the rendered page HTML from
+    the WARC files.
+
+    NOTE: This class is designed to be used via a context manager, e.g.:
+        with WACZClient(<path_to_wacz_file>) as wacz_client:
+            html_content = wacz_client.get_website_content_by_url("http://example.com")
+
+    This guarantees that any open ZipFile file objects will be closed upon context manager
+    exiting.
+    """
+
+    PAGES_FILEPATHS = ("pages/pages.jsonl", "pages/extraPages.jsonl")
+    CDX_INDEX_FILEPATH = "indexes/index.cdx.gz"
+    WARC_DIR = "archive"
+
+    def __init__(self, wacz_filepath: str):
+        self.wacz_filepath = wacz_filepath
+        self._via_context_manager = False
+        self._wacz_archive: zipfile.ZipFile | None = None
+        self._html_websites_df: pd.DataFrame | None = None
+
+    def __enter__(self) -> "WACZClient":
+        """Enter method for use when class instantiated via a context manager.
+
+        This sets the private attribute self._via_context_manager to True, allowing any
+        methods to know if operating with a context manager.
+        """
+        self._via_context_manager = True
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> None:
+        """Exit method for use when class instantiated via a context manager.
+
+        When the context manager closes, this method closes any open file objects from the
+        WACZ ZipFile.
+        """
+        self.close()
+
+    def close(self) -> None:
+        """Close open WACZ ZipFile."""
+        if self._wacz_archive:
+            logger.debug("Closing WACZ ZipFile file handle.")
+            self._wacz_archive.close()
+            self._wacz_archive = None
+
+    @property
+    def wacz_archive(self) -> zipfile.ZipFile:
+        """Property that provides an open ZipFile file object of the target WACZ file.
+
+        NOTE: when accessed, this property will throw a ContextManagerRequiredError
+        exception if not operating in a context manager context.
+        """
+        if not self._via_context_manager:
+            msg = (
+                "Please use this class via a context manager for operations that read "
+                "the WACZ file"
+            )
+            raise ContextManagerRequiredError(msg)
+        if self._wacz_archive is None:
+            with smart_open.open(self.wacz_filepath, "rb") as file_obj:
+                _wacz_archive = zipfile.ZipFile(io.BytesIO(file_obj.read()))
+                self._wacz_archive = _wacz_archive
+        return self._wacz_archive
+
+    def _get_archive_file_obj(self, archive_filepath: str) -> IO[bytes]:
+        """Return a file-like object for a filepath in the WACZ ZipFile."""
+        try:
+            return self.wacz_archive.open(archive_filepath)
+        except KeyError as exc:
```

Comment:
Minor but we've always done `except KeyError as e`

Reply:
Nice catch, think that was some copy pasta.